src/units/abilities.cpp

```diff
@@ -1168,52 +1168,137 @@ void attack_type::modified_attacks(unsigned & min_attacks,
 	}
 }
 
-//Functions used for change damage_type list with damage
-static std::optional<std::string> select_damage_type(const unit_ability_list& abil_list, const std::string& type)
+//functions for select damage type who must be returned for each attribute used in damage_type()
+static std::string select_replacement_type(const unit_ability_list& damage_type_list)
 {
+	std::unordered_map<std::string, unsigned int> type_count;
+	unsigned int max = 0;
+	for(auto& i : damage_type_list) {
+		const config& c = *i.ability_cfg;
+		if(c.has_attribute("replacement_type")) {
+			std::string type = c["replacement_type"].str();
+			unsigned int count = ++type_count[type];
+			if((count > max)) {
+				max = count;
+			}
+		}
+	}
+
+	if (type_count.empty()) return "";
+
 	std::vector<std::string> type_list;
-	for(auto& i : abil_list) {
-		if(!(*i.ability_cfg)[type].str().empty()){
-			type_list.push_back((*i.ability_cfg)[type].str());
+	for(auto& i : type_count){
+		if(i.second == max){
+			type_list.push_back(i.first);
 		}
 	}
-	if(type_list.size() >= 2){
+
+	if(type_list.size() > 1){
```

Comment:
If you think this is a useful optimization then make it an early return. That makes the remaining code easier to read since you can don't need to keep this test in mind for the rest of the function.

The rest of the function could just be:
```cpp
if(type_list.size() == 1) return type_list.front();

std::sort(type_list.begin(), type_list.end());
return type_list.front();
```

`type_list` cannot be empty at this point but if you must you can also add an early return for the empty case.

Reply:
done

> If you think this is a useful optimization then make it an early return. That makes the remaining code easier to read since you can don't need to keep this test in mind for the rest of the function.
> 
> The rest of the function could just be:
> 
> ```c++
> if(type_list.size() == 1) return type_list.front();
> 
> std::sort(type_list.begin(), type_list.end());
> return type_list.front();
> ```
> 
> `type_list` cannot be empty at this point but if you must you can also add an early return for the empty case.

it's done